File: 8.StringtoInteger(atoi).java
class Solution {
  public int myAtoi(String str) {
    str = str.trim();

    if (str == null || str.length() == 0) {
      return 0;
    }

    int start = 0;
    // "result" needs to be double
    double result = 0;
    boolean isNegative = false;

    if (str.charAt(0) == '+') {
      start++;
    } else if (str.charAt(0) == '-') {
      start++;
      isNegative = true;
    }

    for (int i = start; i < str.length(); i++) {
      if (str.charAt(i) < '0' || str.charAt(i) > '9') {
        break;
      }

      // Minus '0' from the ASCII value to get the number
      int digitValue = (int) (str.charAt(i) - '0');
      result = result * 10 + digitValue;
    }

    if (isNegative) {
      result = -result;
    }

    if (result > Integer.MAX_VALUE) {
      return Integer.MAX_VALUE;
    }

    if (result < Integer.MIN_VALUE) {
      return Integer.MIN_VALUE;
    }

    return (int) result;
  }
}

File: 8.StringtoInteger(atoi).js
var myAtoi = function(str) {
  str = str.trim();

  if (str === null || str.length === 0) {
    return 0;
  }

  let result = 0,
    startIndex = 0,
    isNegative = false;

  if (str[0] === '+') {
    startIndex++;
  } else if (str[0] === '-') {
    startIndex++;
    isNegative = true;
  }

  for (let i = startIndex; i < str.length; i++) {
    if (str.codePointAt(i) < 48 || str.codePointAt(i) > 57) {
      break;
    }

    let digitValue = parseInt(String.fromCodePoint(str.codePointAt(i)));
    result = parseInt(result * 10) + digitValue;

    if (result > 2147483647 && isNegative === false) {
      return 2147483647;
    } else if (result > 2147483648 && isNegative) {
      return -2147483648;
    }
  }

  if (isNegative) {
    result = -result;
  }

  return result;
};